### PR TITLE
Ensure RawPath match Path when resolving reference

### LINKED ISSUE #9764: Relative package URLs are not properly resolved when repo URL contains escaped parts
When the repo url contains URL-encoded parts, and doesn't end with `/`, relative downloads are not properly resolved.

Example (with a [patched Gitlab](https://gitlab.com/gitlab-org/gitlab/-/issues/18997#note_530270349)):

```shell
$ helm repo add foo https://gitlab.example.org/api/v4/projects/external-packages%2fargocd/packages/helm/stable
"foo" has been added to your repositories

$ helm show chart foo/argo-cd --debug
show.go:173: [debug] Original chart version: ""
Error: failed to fetch https://gitlab.example.org/api/v4/projects/external-packages/argocd/packages/helm/stable/charts/argo-cd-3.6.5.tgz : 404 Not Found
helm.go:81: [debug] failed to fetch https://gitlab.example.org/api/v4/projects/external-packages/argocd/packages/helm/stable/charts/argo-cd-3.6.5.tgz : 404 Not Found
helm.sh/helm/v3/pkg/getter.(*HTTPGetter).get
        /home/circleci/helm.sh/helm/pkg/getter/httpgetter.go:73
helm.sh/helm/v3/pkg/getter.(*HTTPGetter).Get
        /home/circleci/helm.sh/helm/pkg/getter/httpgetter.go:41
helm.sh/helm/v3/pkg/downloader.(*ChartDownloader).DownloadTo
        /home/circleci/helm.sh/helm/pkg/downloader/chart_downloader.go:99
helm.sh/helm/v3/pkg/action.(*ChartPathOptions).LocateChart
        /home/circleci/helm.sh/helm/pkg/action/install.go:675
main.runShow
        /home/circleci/helm.sh/helm/cmd/helm/show.go:179
main.newShowCmd.func4
        /home/circleci/helm.sh/helm/cmd/helm/show.go:116
github.com/spf13/cobra.(*Command).execute
        /go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:850
github.com/spf13/cobra.(*Command).ExecuteC
        /go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:958
github.com/spf13/cobra.(*Command).Execute
        /go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:895
main.main
        /home/circleci/helm.sh/helm/cmd/helm/helm.go:80
runtime.main
        /usr/local/go/src/runtime/proc.go:204
runtime.goexit
        /usr/local/go/src/runtime/asm_amd64.s:1374
```

As you can see, `%2f` was replaced by `/` which leads to `404 Not found`.

Workaround: Add the slash when issuing `helm repo add`:

```shell
$ helm repo add foo  https://gitlab.k8s.qt.insee.test/api/v4/projects/external-packages%2fargocd/packages/helm/stable/
"foo" has been added to your repositories

$ helm show chart foo/argo-cd
apiVersion: v2
appVersion: 2.0.3
dependencies:
- condition: redis-ha.enabled
  name: redis-ha
  repository: https://dandydeveloper.github.io/charts/
  version: 4.12.14
description: A Helm chart for ArgoCD, a declarative, GitOps continuous delivery tool for Kubernetes.
home: https://github.com/argoproj/argo-helm
icon: https://raw.githubusercontent.com/argoproj/argo/master/docs/assets/argo.png
keywords:
- argoproj
- argocd
- gitops
maintainers:
- name: alexec
- name: alexmt
- name: jessesuen
- name: seanson
name: argo-cd
version: 3.6.5
```

### PR DESCRIPTION
**Pull request by sathieu**
**What this PR does / why we need it**:

When the repo url contains URL-encoded parts, and doesn't end with `/`, relative downloads are not properly resolved.

Example (with a [patched Gitlab](https://gitlab.com/gitlab-org/gitlab/-/issues/18997#note_530270349)):

```shell
$ helm repo add foo https://gitlab.example.org/api/v4/projects/external-packages%2fargocd/packages/helm/stable
"foo" has been added to your repositories

$ helm show chart foo/argo-cd --debug
show.go:173: [debug] Original chart version: ""
Error: failed to fetch https://gitlab.example.org/api/v4/projects/external-packages/argocd/packages/helm/stable/charts/argo-cd-3.6.5.tgz : 404 Not Found
helm.go:81: [debug] failed to fetch https://gitlab.example.org/api/v4/projects/external-packages/argocd/packages/helm/stable/charts/argo-cd-3.6.5.tgz : 404 Not Found
helm.sh/helm/v3/pkg/getter.(*HTTPGetter).get
        /home/circleci/helm.sh/helm/pkg/getter/httpgetter.go:73
helm.sh/helm/v3/pkg/getter.(*HTTPGetter).Get
        /home/circleci/helm.sh/helm/pkg/getter/httpgetter.go:41
helm.sh/helm/v3/pkg/downloader.(*ChartDownloader).DownloadTo
        /home/circleci/helm.sh/helm/pkg/downloader/chart_downloader.go:99
helm.sh/helm/v3/pkg/action.(*ChartPathOptions).LocateChart
        /home/circleci/helm.sh/helm/pkg/action/install.go:675
main.runShow
        /home/circleci/helm.sh/helm/cmd/helm/show.go:179
main.newShowCmd.func4
        /home/circleci/helm.sh/helm/cmd/helm/show.go:116
github.com/spf13/cobra.(*Command).execute
        /go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:850
github.com/spf13/cobra.(*Command).ExecuteC
        /go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:958
github.com/spf13/cobra.(*Command).Execute
        /go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:895
main.main
        /home/circleci/helm.sh/helm/cmd/helm/helm.go:80
runtime.main
        /usr/local/go/src/runtime/proc.go:204
runtime.goexit
        /usr/local/go/src/runtime/asm_amd64.s:1374
```

As you can see, `%2f` was replaced by `/` which leads to `404 Not found`.

Workaround: Add the slash when issuing `helm repo add`:

```shell
$ helm repo add foo  https://gitlab.k8s.qt.insee.test/api/v4/projects/external-packages%2fargocd/packages/helm/stable/
"foo" has been added to your repositories

$ helm show chart foo/argo-cd
apiVersion: v2
appVersion: 2.0.3
dependencies:
- condition: redis-ha.enabled
  name: redis-ha
  repository: https://dandydeveloper.github.io/charts/
  version: 4.12.14
description: A Helm chart for ArgoCD, a declarative, GitOps continuous delivery tool for Kubernetes.
home: https://github.com/argoproj/argo-helm
icon: https://raw.githubusercontent.com/argoproj/argo/master/docs/assets/argo.png
keywords:
- argoproj
- argocd
- gitops
maintainers:
- name: alexec
- name: alexmt
- name: jessesuen
- name: seanson
name: argo-cd
version: 3.6.5
```

**Special notes for your reviewer**:

Fixes: #9764.

**If applicable**:
- [ ] this PR contains documentation
- [ ] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

### CODE REVIEW

**Comment by cndoit18** [comment]
PRs must be linked to an existing issue. If none exists, may you please open one explaining the issue and expected outcome?
Please also write tests to test and verify this PR fixes your issue. Thanks.

**Comment by sathieu** [comment]
> 
> 
> PRs must be linked to an existing issue. If none exists, may you please open one explaining the issue and expected outcome?
> Please also write tests to test and verify this PR fixes your issue. Thanks.

Thanks @cndoit18. Issue #9764 filled.

**Comment by sathieu** [comment]
@cndoit18 This PR has tests and it fixes my problem.

File: pkg/repo/chartrepo.go [hunk 1]
```diff
@@ -285,6 +285,7 @@ func ResolveReferenceURL(baseURL, refURL string) (string, error) {
 	}
 
 	// We need a trailing slash for ResolveReference to work, but make sure there isn't already one
+	parsedBaseURL.RawPath = strings.TrimSuffix(parsedBaseURL.RawPath, "/") + "/"
 	parsedBaseURL.Path = strings.TrimSuffix(parsedBaseURL.Path, "/") + "/"
```
Comment (cndoit18) [inline pkg/repo/chartrepo.go:289]:
```suggestion
```

I would prefer to remove this and change line 277.
```golang
// We need a trailing slash for ResolveReference to work, but make sure there isn't already one
	parsedBaseURL, err := url.Parse(strings.TrimSuffix(baseURL, "/") + "/"
```

**Review by cndoit18 — CHANGES_REQUESTED** [review]

File: pkg/repo/chartrepo.go (hunk 1 above)
Comment (sathieu) [reply pkg/repo/chartrepo.go]: Good idea @cndoit18. Done.

**Review by cndoit18 — APPROVED** [review]
LGTM

**Comment by sathieu** [comment]
@cndoit18 Anything to do on this PR?

**Comment by sathieu** [comment]
Note: There is a similar problem for `helm dependency build` [here](https://github.com/helm/helm/blob/699ea6dcef302d7ded608bc8a1ab6c404de81685/pkg/downloader/manager.go#L780). Should I create another PR or push to this one?

**Comment by sathieu** [comment]
@cndoit18 I went ahead and fixed the following commands:

- `helm install` when `repo_url` is not ending with `/`
- `helm dependency build`
- `helm get` (`ChartDownloader`)
- `helm repo index`

Please review again. I can split in 4 PRs if needed.

**Comment by cndoit18** [comment]
> @cndoit18 I went ahead and fixed the following commands:
> 
> * `helm install` when `repo_url` is not ending with `/`
> * `helm dependency build`
> * `helm get` (`ChartDownloader`)
> * `helm repo index`
> 
> Please review again. I can split in 4 PRs if needed.



> split

I tend to split into multiple PRs

**Comment by sathieu** [comment]
> > @cndoit18 I went ahead and fixed the following commands:
> > 
> > * `helm install` when `repo_url` is not ending with `/`
> > * `helm dependency build`
> > * `helm get` (`ChartDownloader`)
> > * `helm repo index`
> > 
> > Please review again. I can split in 4 PRs if needed.
> 
> I tend to split into multiple PRs

Splitted as:

- #9763 + #9824: When repo URL contains encoded parts and doesn't end with slash (workaround: add trailing `/`)
- #9822: To fix ` helm dependency build`. :fire: This one has **no** workaround
- #9825: To fix `helm repo index`. Edge case.

**Review by technosophos — APPROVED** [review]
LGTM

**Comment by mattfarina** [comment]
For some reason CircleCI does not run on this PR. I created #9922 to try to run tests with the same commits and it has not run yet. There appears to be a partial outage... https://status.circleci.com/incidents/kxw3dyhykqkb

The webhook for this PR is consistently returning a 400 status code (and re-runs do the same thing).

Locally the commits pass tests for me. If not resolved before the next meeting we can talk about this CI issue there.

**Review by mattfarina — APPROVED** [review]

**Comment by mattfarina** [comment]
Tests pass for this. See #9922 for the passing tests. Merging.

**Comment by sathieu** [comment]
Thanks @mattfarina! Would you also take a look at #9822?
